### PR TITLE
ActionController::UnknownFormat should return HTTP 406

### PR DESCRIPTION
**Pull request by maclover7**
See #20666 for details. cc @chancancode

### CODE REVIEW

**Comment by chancancode** [comment]
Thanks for tackling this @maclover7!

From the issue:

> 1. If an action has no templates defined at all, AND it has no respond_to { ... } block, then it should do a 204
> 2. If an action has certain templates defined, AND it has no respond_to { ... } block, then it should do a 406 for formats with no templates
> 3. If an action has a respond_to { ... } block, then it should do a 406 for formats not in the list

Are all three of them fixed now, or only some of them? Can you point me to the test cases that demonstrates these paths or add new ones for each?

**Comment by jmbejar** [comment]
I was also working on this issue but I have something that is not ready for a PR yet [here](https://github.com/jmbejar/rails/commit/8e4eec328748966de195fb2fb5afe15eaf009034). I didn't have enough time to test properly and tests cases has to be completed, however I think it's working as @chancancode proposed.

@maclover7 Are you going to continue working on this issue? If so, feel free to continue in your own and just take my code if that helps. Otherwise, I would open a separate PR asap.

@chancancode Although test cases are not complete yet in my commit, maybe you can confirm that overall direction is what you're expecting?

**Comment by maclover7** [comment]
Hi @jmbejar! I am planning on getting back to this -- my weekend was just a bit hectic.

**Comment by jmbejar** [comment]
@maclover7 Perfect!

**Comment by maclover7** [comment]
@chancancode- updated PR to reflect desired behavior. The three test methods in ` actionpack/test/controller/render_test.rb` correspond in order to the desired behavior numbers in your comment. Let me know if there's anything I can improve upon. :)

File: railties/lib/rails/generators/rails/scaffold_controller/templates/api_controller.rb [hunk 1]
```diff
@@ -41,6 +41,7 @@ def update
   # DELETE <%= route_url %>/1
   def destroy
     @<%= orm_instance.destroy %>
+    render json: {}, status: :no_content
```
Comment (chancancode) [inline railties/lib/rails/generators/rails/scaffold_controller/templates/api_controller.rb]: Why do we need this? Shouldn't the rule...

> If an action has no templates defined at all, AND it has no respond_to { ... } block, then it should do a 204

...already cover this case?

File: actionpack/test/controller/render_test.rb [hunk 2]
```diff
@@ -481,6 +490,21 @@ def test_implicit_no_content_response
     get :empty_action
     assert_response :no_content
   end
+
+  def test_implicit_render_template_no_respond_to
+    get :json_action, format: 'json'
+    assert_response :success
```
Comment (chancancode) [inline actionpack/test/controller/render_test.rb]: Can you put something in the template and assert the content is what you expect?

File: actionpack/test/controller/render_test.rb [hunk 3]
```diff
@@ -481,6 +490,21 @@ def test_implicit_no_content_response
...
+
+    assert_raises ActionController::UnknownFormat do
```
Comment (chancancode) [inline actionpack/test/controller/render_test.rb]: Can you split this into a different test case?

File: actionview/lib/action_view/lookup_context.rb [hunk 4]
```diff
@@ -132,6 +132,11 @@ def exists?(name, prefixes = [], partial = false, keys = [], **options)
       end
       alias :template_exists? :exists?
 
+      def template_exists_in_any_format?(name, prefixes = [], partial = false, keys = [], **options)
+        options_with_all_formats = options.merge(formats: default_formats)
+        @view_paths.exists?(*args_for_lookup(name, prefixes, partial, keys, options_with_all_formats))
+      end
```
Comment (chancancode) [inline actionview/lib/action_view/lookup_context.rb]: Should this be private? (I don't know what the rules are for the `ViewPaths` class. If the class itself is already private then it's probably fine.

Also, can you add a comment for this method? I assume you probably know more about the differences between these methods (particularly `template_exists_in_any_format?` vs `template_exists?`), so if you can document both of them for future readers that would be great

Comment (chancancode) [inline actionview/lib/action_view/lookup_context.rb]: Particularly, if it's checking if *any* template exists for this name, I don't know why we need to pass in the variant.

Comment (chancancode) [inline actionview/lib/action_view/lookup_context.rb]: I also don't know what the semantics should be for variants, do you have thoughts about this?

(It also seems good to enumerate the rules in a comment somewhere, by the way: https://github.com/rails/rails/issues/20666#issuecomment-116882771)

File: actionview/test/template/lookup_context_test.rb [hunk 5]
```diff
@@ -206,6 +206,16 @@ def teardown
     @lookup_context.prefixes = ["foo"]
     assert_equal ["foo"], @lookup_context.prefixes
   end
+
+  test "checks for any template with name" do
+    # returns false if cannot be found
+    result = @lookup_context.template_exists_in_any_format?('foo', ['foo'], false, [])
+    assert_equal(false, result)
```
Comment (chancancode) [inline actionview/test/template/lookup_context_test.rb]: assert_not

File: actionview/test/template/lookup_context_test.rb [hunk 6]
```diff
@@ -206,6 +206,16 @@ def teardown
...
+
+    # returns true if can be found
+    result = @lookup_context.template_exists_in_any_format?('hello_world', ['test'], false, [])
+    assert_equal(true, result)
```
Comment (chancancode) [inline actionview/test/template/lookup_context_test.rb]: assert

File: actionview/test/template/lookup_context_test.rb [hunk 7]
```diff
@@ -206,6 +206,16 @@ def teardown
     @lookup_context.prefixes = ["foo"]
     assert_equal ["foo"], @lookup_context.prefixes
   end
+
+  test "checks for any template with name" do
+    # returns false if cannot be found
```
Comment (chancancode) [inline actionview/test/template/lookup_context_test.rb]: This does not explain anything, so either :scissors: it or elaborate

File: actionview/test/template/lookup_context_test.rb [hunk 8]
```diff
@@ -206,6 +206,16 @@ def teardown
...
+    result = @lookup_context.template_exists_in_any_format?('foo', ['foo'], false, [])
+    assert_equal(false, result)
+
+    # returns true if can be found
```
Comment (chancancode) [inline actionview/test/template/lookup_context_test.rb]: ditto

File: railties/lib/rails/generators/rails/scaffold_controller/templates/api_controller.rb (hunk 1 above)
Comment (maclover7) [inline railties/lib/rails/generators/rails/scaffold_controller/templates/api_controller.rb]: I remember that one of the tests was failing when I originally pushed, but I think that's because I changed the default to render a 406 if no content was provided. Removed the line and pushed to Travis, let's see how it turns up!

File: actionview/lib/action_view/lookup_context.rb (hunk 4 above)
Comment (maclover7) [inline actionview/lib/action_view/lookup_context.rb]: I think I'm just passing in arguments required by ViewPaths. I figured that if the method might be reused elsewhere, might as well provide correct options. Don't think we'll be able to make this a private method, because we need to access this `ActionView` method from `ActionPack`.

**Comment by maclover7** [comment]
@chancancode updated

File: actionpack/lib/action_controller/metal/implicit_render.rb [hunk 9]
```diff
@@ -21,8 +21,14 @@ def default_render(*args)
         if block_given?
           yield(*args)
         else
-          logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content" if logger
-          super
+          if no_templates?
+            logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content" if logger
+            super
+          else
+            format = request.format.to_sym.upcase if request.format
```
Comment (chancancode) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: Why `to_sym`?

Comment (chancancode) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: Also, what is the case when `request.format` is falsey and what happens in the log message below in that case?

Comment (maclover7) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: The chances that `request.format` is falsey is pretty unlikely -- I think a bunch of other things within AP would probably break if this happens. The reason why we have to perform some cleaning functions on the `request.format` is because it's a `Mime::Type`. Changed the behavior a little to make it more understandable, and in the syntax style of `respond_to`.

File: actionview/lib/action_view/lookup_context.rb (hunk 4 above)
Comment (jmbejar) [inline actionview/lib/action_view/lookup_context.rb]: :+1: to add more comments and maybe write the rules (https://github.com/rails/rails/issues/20666#issuecomment-116882771) somewhere explicitly in the code.

Regarding variants, I think the responsibility of this method should be to check if any template exists for a given `name` and `variant`. For each possible pair of ("template name", "variant"), I can easily apply @chancancode's rules to determine whether is the HTTP code, 406 or 204, for the response.

Does not it make sense for you, guys?

**Comment by maclover7** [comment]
@chancancode Updated PR with more documentation + removed arguments that were not needed.

File: actionpack/lib/action_controller/metal/implicit_render.rb [hunk 10]
```diff
@@ -32,5 +44,13 @@ def method_for_action(action_name)
         "default_render"
       end
     end
+
+    private
+      def no_templates?
+        # Check and see if any templates exist for the given action name, and prefixes.
+        # Variants are provided also as an argument, to check in case a template exists
+        # for a format not specifically declared in the `respond_to` block.
+        !template_exists_in_any_format?(action_name.to_s, _prefixes, { variants: request.variant })
```
Comment (chancancode) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: I'm still not sure why we need to pass the variant here; isn't the point of `template_exists_in_any_format?` to return true if *any* template (for any formats or variant or not) exists? Why are we taking into account the variant in the lookup? Is this a technical limitation of the architecture?

By the way, does this work with variants? Probably good to add some tests for that too

File: actionpack/lib/action_controller/metal/implicit_render.rb [hunk 11]
```diff
@@ -21,8 +21,20 @@ def default_render(*args)
...
+            # If an action has no templates defined at all, and has no `respond_to` block,
+            # then `head :no_content` will be returned
+            logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content" if logger
+            super
+          else
+            # If an action has certain templates defined, and has no `respond_to` block,
+            # then `head :not_acceptable` will be returned for formats with no templates
+            #
+            # If an action has a `respond_to` block, `head :not_acceptable` will be returned for formats that aren't specified
+            format = request.format.symbol.to_s.delete(':') # Make request format more readable
```
Comment (chancancode) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: Why is this necessary?

```
irb(main):001:0> :zomg.to_s
=> "zomg"
```

File: actionpack/test/controller/render_test.rb [hunk 12]
```diff
@@ -481,6 +490,22 @@ def test_implicit_no_content_response
     get :empty_action
     assert_response :no_content
   end
+
+  def test_implicit_render_template_no_respond_to
+    get :xml_action, format: 'atom'
+    assert_equal "<atom>\n</atom>\n", @response.body
+    assert_response :success
+
+    assert_raises ActionController::UnknownFormat do
```
Comment (chancancode) [inline actionpack/test/controller/render_test.rb]: Would prefer to split this into another test case

File: actionpack/lib/action_controller/metal/implicit_render.rb (hunk 10 above)
Comment (maclover7) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: We need to pass in the variant, or else a test like:

```ruby
def test_implicit_render_template_unknown_format_response
    assert_raises ActionController::UnknownFormat do
      get :xml_action, format: 'json'
    end
  end
```

will fail.

Added some additional `AV` tests to check for variants.

Comment (chancancode) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: But do you know *why* it fails?

Comment (chancancode) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: (To be clear – I don't; and I would like to understand that better)

Comment (maclover7) [inline actionpack/lib/action_controller/metal/implicit_render.rb]: It would fail if only the action name and prefix were passed as arguments, because since the format is JSON, it matters that whatever being returned is coming from a JSON template. Otherwise, just the XML (or whatever) would be returned. It's possible that the method name is what's confusing -- I'm all ears to change it if there's a name that's easier to understand.

Also, I tried swapping out the new method in `no_templates?` with the already existing `template_exists?` method, and that caused the test above to fail.

File: actionpack/test/controller/mime/respond_to_test.rb [hunk 13]
```diff
@@ -619,21 +619,25 @@ def test_invalid_variant
     logger = ActiveSupport::LogSubscriber::TestHelper::MockLogger.new
     old_logger, ActionController::Base.logger = ActionController::Base.logger, logger
 
-    get :variant_with_implicit_rendering, params: { v: :invalid }
-    assert_response :no_content
+    assert_raises ActionController::UnknownFormat do
```
Comment (maclover7) [inline actionpack/test/controller/mime/respond_to_test.rb]: These all need to be changed to `ActionController::UnknownFormat` because there is a template defined for this controller action (I could find only 1 template), so this means that an error is raised if the format and/or variant is not recognized.

Comment (chancancode) [inline actionpack/test/controller/mime/respond_to_test.rb]: :+1:

**Comment by chancancode** [comment]
Closed by #23827
